manifest/oci_index.go

```diff
@@ -69,36 +76,60 @@ func (index *OCI1Index) UpdateInstances(updates []ListUpdate) error {
 	return nil
 }
 
-// ChooseInstance parses blob as an oci v1 manifest index, and returns the digest
-// of the image which is appropriate for the current environment.
-func (index *OCI1Index) ChooseInstance(ctx *types.SystemContext) (digest.Digest, error) {
-	wantedPlatforms, err := platform.WantedPlatforms(ctx)
-	if err != nil {
-		return "", fmt.Errorf("getting platform information %#v: %w", ctx, err)
-	}
-	for _, wantedPlatform := range wantedPlatforms {
-		for _, d := range index.Manifests {
+func instanceBetterMatchesPlatforms(ctx *types.SystemContext, manifests []imgspecv1.Descriptor, platformCheck bool, wantedPlatform imgspecv1.Platform) digest.Digest {
+	var matchedDigest digest.Digest
+	matchedDigest = ""
+	for _, d := range manifests {
```

Comment:
This structure feels _very_ unnatural to me.

There is now a loop in  `ChooseInstance`; that loop terminates on a first match, and calls `instanceBetterMatchesPlatform`, which initiates _another_ loop, and always finds an image, and returns it, and that terminates `ChooseIntance`, AFAICS. Why have two loops at all?

The specific check for `GivePriority`, and the _immediate_ return in that case, also doesn’t generalize to the future at all. We might have more compression formats.

---

The suggestion with `instanceBetterMatchesPlatforms` was to define it as a predicate that defines an ordering on platforms; and then `ChooseInstance` can just find a minimum = best match.

Conceptually:
```swift
func ChooseInstance(ctx: SystemContext) -> Digest {
    let wantedPlatforms = platform.WantedPlatforms(ctx: ctx)
    let relevant = index.Manifests.filter({ manifest in
        platformMatches(manifest: manifest, wantedPlatforms: wantedPlatforms)
    })
    let best := relevant.min(by: { x, y in instanceComparator(x, y, platformOrder: wantedPlatforms) == .orderedAscending })
    return best?.Digest
}

func instanceComparator(x, y Descriptor, platformOrder: …) -> ComparisonResult {
    if let v = platformComparator(x, y, orderingBy: platformOrder); v != .orderedSame { return v }
    if let v = compressionComparator(x, y, preferenceBasedOn: compressionPreferenceOptions); v != .orderedSame { return v }
    return .orderedSame
}
```

Now, the actual implementation would probably look a bit different — e.g. finding out the platform ordering means a linear search through `wantedPlatforms`, so we would only want to do that once per item, instead of deep in the `instanceComparator`, but the _rough idea_ was to separate the “is this a better match” and the iteration code.

Reply:
@mtrmac I have modified code so no unnecessary loop is added and original code is kept intact, I like the idea of adding comparator but while I was implementing that I saw it requires us to add additional sorting infra while our feature set of `ChooseInstance` and `copy.Options` does not have choices it only supports `zstd` or `gzip` so code for doing priority selection looked early optimization with additional sorting infrastructure, hence I refrained from adding that and was thinking to add that complexity when one more choice arise. 

But anyways I am okay with whatever you prefer so could you please take a look a new push and let me know ?

Reply:
We don’t really need _sorting_, `min` is just:
```go
var best *Manifest
for _, m := range manifests {
    if /* manifest is acceptable at all */ && (best == nil || instanceBetterMatchesPlatforms(manifest, best) {
        best = manifest
    }
}
```

Reply:
… I’ll take a more detailed look in a few hours.